tests/unit/utils/test_sqlalchemy_exec_utils.py

```diff
@@ -62,13 +62,45 @@ def execute(self, query):
 ERROR = OperationalError("A bad thing happend", {}, None)
 
 
+def test_execute_with_retry():
+    tracker = {"error": True, "finished": []}
+
+    def log(msg):
+        tracker["logs"] = tracker.get("logs", []) + [msg]
+
+    num_parts = 4
+
+    def execute():  # Return an iterator to mimic Connection.execute
+        # Succeed for 0, 1, fail for 2 on the first go, then succeed for 2, 3
+        def execute_part(part, tr):
+            if "error" in tr and part == 2:
+                tr.pop("error")
+                raise ERROR
+            return part
+
+        return (execute_part(part, tracker) for part in range(num_parts))
+
+    execute_with_retry = execute_with_retry_factory(
+        execute, max_retries=3, retry_sleep=0, backoff_factor=2, log=log
+    )
+
+    # Do something with the results
+    for part in execute_with_retry():
+        tracker["finished"].append(part)
+
+    assert tracker["finished"] == list(range(num_parts))
+    assert "Retrying query (attempt 1 / 3)" in tracker["logs"]
```

Comment:
Asserting on the log message too is a nice addition. I think these were untested before so thanks for including this.

You can do this slightly more simply though like this:
```py
log_messages = []

execute_with_retry = execute_with_retry_factory(
    ...    
    log=log_messages.append,
)